### PR TITLE
style: Move display outside of mako

### LINKED ISSUE #19658: Move display outside of mako.
I'll take this one as a sub-issue of #19015

### PR DESCRIPTION
**Pull request by CYBAI**
This is a sub-PR of #19015 
r? emilio 

---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #19658 
- [x] These changes do not require tests

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/19659)

### CODE REVIEW

File: components/style/values/specified/box.rs [hunk 1]
```diff
@@ -7,15 +7,429 @@
...
+    pub fn cascade_property_custom(
+        _declaration: &PropertyDeclaration,
+        context: &mut Context
+    ) {
+        longhands::_servo_display_for_hypothetical_box::derive_from_display(context);
+        longhands::_servo_text_decorations_in_effect::derive_from_display(context);
+    }
+}
+
+impl Parse for Display {
```
Comment (emilio) [inline components/style/values/specified/box.rs]: You should be able to `derive` this, shouldn't you?

File: components/style/values/specified/box.rs [hunk 2]
```diff
@@ -7,15 +7,429 @@
...
 use values::KeyframesName;
+#[cfg(feature = "servo")]
+use values::computed::Context;
 use values::generics::box_::AnimationIterationCount as GenericAnimationIterationCount;
 use values::generics::box_::VerticalAlign as GenericVerticalAlign;
 use values::specified::{AllowQuirks, Number};
 use values::specified::length::LengthOrPercentage;
 
+// FIXME(emilio): Why does this reinvent the wheel, again?
+#[allow(non_camel_case_types)]
```
Comment (emilio) [inline components/style/values/specified/box.rs]: Why is this needed?

**Review by emilio — COMMENTED** [review]
Thanks for doing this one :)

File: components/style/values/specified/box.rs (hunk 2 above)
Comment (CYBAI) [reply components/style/values/specified/box.rs]: I think after moving this `enum` into specified module, we don't need this anymore.
Besides, I think even in the old way, maybe we don't need it either?

File: components/style/values/specified/box.rs (hunk 1 above)
Comment (CYBAI) [reply components/style/values/specified/box.rs]: Ooooooh, yeah, I think I should try to use `derive` for it.
However, I have only one concern to make it use `derive`.
For `Display::Flex` and `Display::InlineFlex`, they are originally matches with `-webkit-` prefix.
If we try to derive `Parse`, I think it will be matched without the prefix?

Comment (CYBAI) [reply components/style/values/specified/box.rs]: Just update the PR with `derive` `Parse`!
I'd like to know if there's any test error with it :)

Comment (emilio) [reply components/style/values/specified/box.rs]: Oh, that will definitely cause tests to fail. I opened  #19665 with a solution for that.

File: components/style/values/specified/box.rs (hunk 2 above)
Comment (emilio) [reply components/style/values/specified/box.rs]: Yeah, probably just leftover.

Comment (emilio) [reply components/style/values/specified/box.rs]: And now it uses `derive(Parse)` the `FIXME` can go away too :)

File: components/style/values/specified/box.rs [hunk 3]
```diff
@@ -7,15 +7,277 @@
...
+    FlowRoot,
+    #[cfg(feature = "gecko")]
+    /// `box` variant for `webkit`
+    WebkitBox,
+    #[cfg(feature = "gecko")]
+    /// `inline-box` variant for `webkit`
+    WebkitInlineBox,
+    #[cfg(feature = "gecko")]
+    /// `box` variant for `moz`
+    MozBox,
```
Comment (emilio) [inline components/style/values/specified/box.rs]: I'd just `#[allow(missing_docs)]`, for the variants, and remove the comments, they are more noise than good I think, and for keywords that's just fine. In particular, these for example are wrong (`-moz-box` is a XUL display type, as lots of the ones below starting with `-moz`).

File: components/style/properties/longhand/box.mako.rs [hunk 4]
```diff
@@ -13,229 +13,41 @@
...
+    if product == "gecko":
+        display_values += """grid inline-grid ruby ruby-base ruby-base-container
+            ruby-text ruby-text-container contents flow-root -webkit-box
+            -webkit-inline-box -moz-box -moz-inline-box -moz-grid -moz-inline-grid
+            -moz-grid-group -moz-grid-line -moz-stack -moz-inline-stack -moz-deck
+            -moz-popup -moz-groupbox""".split()
+%>
 
-</%helpers:longhand>
+${helpers.gecko_keyword_conversion(
```
Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: It's very unfortunate that we still need to list all the values here... Can you double-check that adding a new value to the `Display` enum at least causes a compile error here?

And can you leave a `FIXME` above like:

```python
# FIXME(emilio): Listing all the display values here is very unfortunate, we should teach C++ to use the
# Rust enum directly, or generate the conversions to `StyleDisplay`.
```

File: components/style/values/specified/box.rs (hunk 1 above)
Comment (CYBAI) [reply components/style/values/specified/box.rs]: WOW! Great solution! Let me try to fix it!

File: components/style/properties/longhand/box.mako.rs (hunk 4 above)
Comment (CYBAI) [reply components/style/properties/longhand/box.mako.rs]: Sure, I'll double-check the `Display` enum again and add this `FIXME` comment here.

File: components/style/values/specified/box.rs (hunk 1 above)
Comment (CYBAI) [reply components/style/values/specified/box.rs]: By the way, the [CI build](https://travis-ci.org/servo/servo/builds/323504005) in the patch without `#[parse(alias = "-webkit-flex")]` still succeeded.
I guess it didn't be tested. Shall we have another PR to add test for it?

File: components/style/values/specified/box.rs (hunk 3 above)
Comment (CYBAI) [reply components/style/values/specified/box.rs]: Using `#[allow(missing_docs)]` makes this `enum` more clear :P

File: components/style/properties/longhand/box.mako.rs (hunk 4 above)
Comment (CYBAI) [reply components/style/properties/longhand/box.mako.rs]: Yeah, I tried to add a `display-test` into `display_values` and it caused compilation error with 

```diff
diff --git a/components/style/properties/longhand/box.mako.rs b/components/style/properties/longhand/box.mako.rs
index 641d72afae..b32248b513 100644
--- a/components/style/properties/longhand/box.mako.rs
+++ b/components/style/properties/longhand/box.mako.rs
@@ -32,7 +32,7 @@ ${helpers.predefined_type(
     display_values = """inline block inline-block
     table inline-table table-row-group table-header-group table-footer-group
     table-row table-column-group table-column table-cell table-caption
-    list-item none
+    list-item none display-test
     """.split()
     webkit_prefixed_display_values = "flex inline-flex".split()
     display_values += webkit_prefixed_display_values
```

```rust
   Compiling style v0.0.1 (file:///Users/cybai/codespace/servo/components/style)
error[E0599]: no associated item named `DisplayTest` found for type `gecko_bindings::structs::root::mozilla::StyleDisplay` in the current scope
    --> servo/target/geckolib/debug/build/style-55046d2555925070/out/properties.rs:7989:23
     |
7989 |                     = structs::StyleDisplay::DisplayTest as u32;
     |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

error: Could not compile `style`.

To learn more, run the command again with --verbose.
GeckoLib build completed in 0:00:18
```

File: components/style/properties/longhand/box.mako.rs [hunk 5]
```diff
@@ -13,229 +13,42 @@
...
+)}
 
-</%helpers:longhand>
+// FIXME(emilio): Listing all the display values here is very unfortunate, we should teach C++ to use the
+// Rust enum directly, or generate the conversions to `StyleDisplay`.
+<%
+    display_values = """inline block inline-block
+    table inline-table table-row-group table-header-group table-footer-group
+    table-row table-column-group table-column table-cell table-caption
+    list-item none
```
Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: This is Gecko-only code now, so no need to ad the `if product == "gecko":`, nor special-casing `webkit_prefixed_display_values`, just put them on the same list.

File: components/style/values/specified/box.rs [hunk 6]
```diff
@@ -7,15 +7,194 @@
...
+pub enum Display {
+    Inline, Block, InlineBlock,
+    Table, InlineTable, TableRowGroup, TableHeaderGroup,
+    TableFooterGroup, TableRow, TableColumnGroup,
+    TableColumn, TableCell, TableCaption, ListItem, None,
+    #[parse(aliases = "-webkit-flex")]
+    Flex,
+    #[parse(aliases = "-webkit-inline-flex")]
+    InlineFlex,
+    #[cfg(feature = "gecko")]
```
Comment (emilio) [inline components/style/values/specified/box.rs]: I don't think this works as you think, you still need a `cfg` per variant, let's put them back one per line.

File: components/style/values/specified/box.rs [hunk 7]
```diff
@@ -7,15 +7,194 @@
...
+
+    /// Returns whether an element with this display type is a line
+    /// participant, which means it may lay its children on the same
+    /// line as itself.
+    pub fn is_line_participant(&self) -> bool {
+        match *self {
+            Display::Inline => true,
+            #[cfg(feature = "gecko")]
+            Display::Contents | Display::Ruby
+                | Display::RubyBaseContainer => true,
```
Comment (emilio) [inline components/style/values/specified/box.rs]: nit: indentation here looks somewhat weird, what about:

```rust
    pub fn is_line_participant(&self) -> bool {
        match *self {
            Display::Inline => true,
            #[cfg(feature = "gecko")]
            Display::Contents |
            Display::Ruby |
            Display::RubyBaseContainer => true,
```

Not particularly sold on it, so if you think it's cleaner as-is that's fine for me too.

File: components/style/properties/longhand/box.mako.rs [hunk 8]
```diff
@@ -13,229 +13,42 @@
...
-    ${helpers.gecko_keyword_conversion(Keyword('display', ' '.join(values),
-                                               gecko_enum_prefix='StyleDisplay',
-                                               gecko_strip_moz_prefix=False))}
+${helpers.predefined_type(
+    "display",
+    "Display",
+    "computed::Display::inline()",
+    initial_specified_value="specified::Display::inline()",
+    animation_value_type="discrete",
+    needs_context=False,
```
Comment (emilio) [inline components/style/properties/longhand/box.mako.rs]: I believe you can remove this line, too.

**Review by emilio — COMMENTED** [review]
Few more changes needed, but looks much nicer!

File: components/style/values/specified/box.rs (hunk 1 above)
Comment (emilio) [reply components/style/values/specified/box.rs]: Sure, the CI build succeeds, but that's only a portion of the tests we run. To run all the tests you should push to Mozilla's try server (https://wiki.mozilla.org/ReleaseEngineering/TryServer), which I think you don't have commit access to because it's a Gecko thingie.

For PRs that just move code around I don't use to push them if I'm confident the changes are correct (we run the tests when the patch lands anyway), but for bigger changes I push and get a link like: https://treeherder.mozilla.org/#/jobs?repo=try&revision=9f764414e0f2341c4494164b8341ab9d901e7524

Comment (CYBAI) [reply components/style/values/specified/box.rs]: Thanks for the information.

File: components/style/values/specified/box.rs (hunk 7 above)
Comment (CYBAI) [reply components/style/values/specified/box.rs]: I was thinking about

```rust
match *self {
  Display::Inline => true,
  #[cfg(feature = "gecko")]
  Display::Contents
    | Display::Ruby
    | Display::RubyBaseContainer  => true,
```

but waiting for your comments :P

I'll follow the indentation you provided above, thanks!.

Comment (emilio) [reply components/style/values/specified/box.rs]: That works for me :)

File: components/style/properties/longhand/box.mako.rs (hunk 5 above)
Comment (CYBAI) [reply components/style/properties/longhand/box.mako.rs]: Yeah! You're right!!
I'll put the string of `display_values` as the second argument of `Keyword` and remove this variable.

File: components/style/properties/longhand/box.mako.rs (hunk 8 above)
Comment (CYBAI) [reply components/style/properties/longhand/box.mako.rs]: It seems we still need this line.

If I remove it, I'll get the error like following:

```rust
error[E0061]: this function takes 1 parameter but 2 parameters were supplied
    --> servo/target/geckolib/debug/build/style-55046d2555925070/out/properties.rs:7874:39
     |
7874 |             specified::Display::parse(context, input)
     |                                       ^^^^^^^^^^^^^^ expected 1 parameter
     |
    ::: components/style/values/specified/box.rs
     |
24   | #[derive(Clone, Copy, Debug, Eq, Hash, MallocSizeOf, Parse, PartialEq, ToComputedValue, ToCss)]
     |                                                      ----- defined here

error: aborting due to previous error
```

File: components/style/values/specified/box.rs (hunk 6 above)
Comment (CYBAI) [reply components/style/values/specified/box.rs]: Just add them back! I think I have misunderstanding to `attributes` 😓

**Comment by bholley** [comment]
I think we should get @CYBAI L1 commit access to mozilla-central to facilitate try pushes. @emilio, do you mind helping that process along?

**Review by emilio — APPROVED** [review]
LGTM, thanks as always @CYBAI!

**Comment by emilio** [comment]
@CYBAI L1 commit access allows you to push to the try server, do you have an account in https://bugzilla.mozilla.org?

**Comment by CYBAI** [comment]
@emilio I don't have an account in bugzilla now. I'll create one later! Thanks.
